File: examples/lh-server/typescript/src/07-child-workflows/main.ts
import { createTaskWorker, LHConfig, Workflow } from "littlehorse-client";
import { z } from "zod";
import { closeOnShutdown } from "../config.js";

function greet(name: string): string {
  const result = `Hello, ${name}!`;
  console.log(result);
  return result;
}

const config = LHConfig.from({});
const worker = createTaskWorker(greet, "greet", config, {
  inputVars: { name: z.string() },
  outputSchema: z.string(),
});

const child = Workflow.newWorkflow("greeting-child", (wf) => {
  const name = wf.declareStr("name").required();
  wf.complete(wf.execute("greet", name));
});

const parent = Workflow.newWorkflow("greeting-parent", (wf) => {
  const inputName = wf.declareStr("input-name").required();
  const childOutput = wf.declareStr("child-output");

  const childWf = wf.runWf("greeting-child", { name: inputName });
  wf.execute("greet", "hi from parent");
  childOutput.assign(wf.waitForChildWf(childWf));
});

await closeOnShutdown([worker]);
await worker.registerTaskDef();
await child.registerWfSpec(config);
await parent.registerWfSpec(config);
console.log('Run with: lhctl run greeting-parent input-name "Obi-Wan"');
await worker.start();

File: examples/lh-server/python/07-child-workflows/main.py
import asyncio
import sys

import littlehorse
from littlehorse.config import LHConfig
from littlehorse.worker import LHTaskWorker
from tasks import greet
from workflow import get_workflows


def register(config: LHConfig) -> None:
    littlehorse.create_task_def(greet, "greet", config)
    for workflow in get_workflows():
        littlehorse.create_workflow_spec(workflow, config)


async def workers(config: LHConfig) -> None:
    await littlehorse.start(LHTaskWorker(greet, "greet", config))


if __name__ == "__main__":
    config = LHConfig()
    if len(sys.argv) != 2 or sys.argv[1] not in {"register", "workers"}:
        raise SystemExit("Usage: main.py [register|workers]")
    register(config) if sys.argv[1] == "register" else asyncio.run(workers(config))
